go/tools/gazelle/merger/merger.go

```diff
@@ -46,112 +46,309 @@ var (
+		} else {
+			genAttr := genRule.Attr(k)
+			var err error
+			mergedAttr, err = mergeExpr(genAttr, oldAttr)
+			if err != nil {
+				// TODO: add a verbose mode and log errors like this.
+				mergedAttr = genAttr
 			}
 		}
-		for _, c := range s.Comment().Before {
-			if strings.HasPrefix(c.Token, gazelleIgnore) {
-				return GazelleIgnoreError{c.Start}
+
+		if mergedAttr != nil {
+			mergedRule.SetAttr(k, mergedAttr)
+		}
+	}
+
+	for _, k := range genRule.AttrKeys() {
+		if mergedRule.Attr(k) == nil {
+			mergedRule.SetAttr(k, genRule.Attr(k))
+		}
+	}
+
+	return &merged
+}
+
+// mergeExpr combines information from gen and old and returns an updated
+// expression. The following kinds of expressions are recognized:
+//
+//   * nil
+//   * strings (can only be merged with strings)
+//   * lists of strings
+//   * a call to select with a dict argument. The dict keys must be strings,
+//     and the values must be lists of strings.
+//   * a list of strings combined with a select call using +. The list must
+//     be the left operand.
+//
+// An error is returned if the expressions can't be merged, for example
+// because they are not in one of the above formats.
+func mergeExpr(gen, old bzl.Expr) (bzl.Expr, error) {
+	if _, ok := gen.(*bzl.StringExpr); ok {
+		if shouldKeep(old) {
+			return old, nil
+		}
+		return gen, nil
+	}
+
+	genList, genDict, err := attrListAndDict(gen)
+	if err != nil {
+		return nil, err
+	}
+	oldList, oldDict, err := attrListAndDict(old)
+	if err != nil {
+		return nil, err
+	}
+
+	mergedList := mergeList(genList, oldList)
+	mergedDict, err := mergeDict(genDict, oldDict)
+	if err != nil {
+		return nil, err
+	}
+
+	var mergedSelect bzl.Expr
+	if mergedDict != nil {
+		mergedSelect = &bzl.CallExpr{
+			X:    &bzl.LiteralExpr{Token: "select"},
+			List: []bzl.Expr{mergedDict},
+		}
+	}
+
+	if mergedList == nil {
+		return mergedSelect, nil
+	}
+	if mergedSelect == nil {
+		return mergedList, nil
+	}
+	return &bzl.BinaryExpr{
+		X:  mergedList,
+		Op: "+",
+		Y:  mergedSelect,
+	}, nil
+}
+
+// attrListAndDict matches an expression and attempts to extract either a list
+// of expressions, a call to select with a dictionary, or both.
+// An error is returned if the expression could not be matched.
+func attrListAndDict(expr bzl.Expr) (*bzl.ListExpr, *bzl.DictExpr, error) {
+	if expr == nil {
+		return nil, nil, nil
+	}
+	switch expr := expr.(type) {
+	case *bzl.ListExpr:
+		return expr, nil, nil
+	case *bzl.CallExpr:
+		if x, ok := expr.X.(*bzl.LiteralExpr); ok && x.Token == "select" && len(expr.List) == 1 {
+			if d, ok := expr.List[0].(*bzl.DictExpr); ok {
+				return nil, d, nil
+			}
+		}
+	case *bzl.BinaryExpr:
+		if expr.Op == "+" {
+			if l, ok := expr.X.(*bzl.ListExpr); ok {
+				if call, ok := expr.Y.(*bzl.CallExpr); ok && len(call.List) == 1 {
+					if x, ok := call.X.(*bzl.LiteralExpr); ok && x.Token == "select" {
+						if d, ok := call.List[0].(*bzl.DictExpr); ok {
+							return l, d, nil
```

Comment:
The problem with flattening (and the reason I did not suggest it) is now if you want to handle a non plus binary expression, you have to un-flatten again, and the more cases you add the worse it gets.
I have hit this a few times doing ast manipulation code in the past, not found a really good answer yet...
The cleanest I have found is to have "matching" functions that do all the tests flattened like  you do now, and return all valid parts, so it's one if per pattern. That generally that leads to inefficiencies where you test the same things in many conditions though, and because they are often nested they never get inlined well, but at least it means everywhere you want to test for the same pattern you are using the same code to do it.